packages/astro/src/content/runtime.ts

```diff
@@ -319,9 +337,84 @@ type RenderResult = {
 	remarkPluginFrontmatter: Record<string, any>;
 };
 
+const CONTENT_LAYER_IMAGE_REGEX = /__ASTRO_IMAGE_="([^"]+)"/g;
+
+async function updateImageReferencesInBody(html: string, fileName: string) {
+	// @ts-expect-error Virtual module
+	const { default: imageAssetMap } = await import('astro:asset-imports');
+
+	const imageObjects = new Map<string, GetImageResult>();
+
+	// @ts-expect-error Virtual module resolved at runtime
+	const { getImage } = await import('astro:assets');
+
+	// First load all the images. This is done outside of the replaceAll
+	// function because getImage is async.
+	for (const [_full, imagePath] of html.matchAll(CONTENT_LAYER_IMAGE_REGEX)) {
+		try {
+			const decodedImagePath = JSON.parse(imagePath.replaceAll('&#x22;', '"'));
+			const id = imageSrcToImportId(decodedImagePath.src, fileName);
+
+			const imported = imageAssetMap.get(id);
+			if (!id || imageObjects.has(id) || !imported) {
+				continue;
+			}
+			const image: GetImageResult = await getImage({ ...decodedImagePath, src: imported });
+			imageObjects.set(imagePath, image);
+		} catch (e) {
+			throw new Error(`Failed to parse image reference: ${imagePath}`);
+		}
+	}
+
+	return html.replaceAll(CONTENT_LAYER_IMAGE_REGEX, (full, imagePath) => {
+		const image = imageObjects.get(imagePath);
+
+		if (!image) {
+			return full;
+		}
+
+		const { index, ...attributes } = image.attributes;
+
+		return Object.entries({
+			...attributes,
+			src: image.src,
+			srcset: image.srcSet.attribute,
+		})
+			.map(([key, value]) => (value ? `${key}=${JSON.stringify(String(value))}` : ''))
+			.join(' ');
+	});
+}
+
+function updateImageReferencesInData<T extends Record<string, unknown>>(
+	data: T,
+	fileName: string,
+	imageAssetMap: Map<string, ImageMetadata>
+): T {
+	return new Traverse(data).map(function (ctx, val) {
+		if (typeof val === 'string' && val.startsWith(IMAGE_IMPORT_PREFIX)) {
+			const src = val.replace(IMAGE_IMPORT_PREFIX, '');
+			const id = imageSrcToImportId(src, fileName);
+			if (!id) {
+				ctx.update(src);
+				return;
+			}
+			const imported = imageAssetMap.get(id);
+			if (imported) {
+				ctx.update(imported);
+			} else {
+				ctx.update(src);
+			}
+		}
+	});
+}
+
 async function renderEntry(entry?: DataEntry) {
-	console.log('entry:', entry?.rendered?.metadata);
-	const Content = createComponent(() => serverRender`${unescapeHTML(entry?.rendered?.html)}`);
+	const html =
+		entry?.rendered?.metadata?.imagePaths?.length && entry.filePath
+			? await updateImageReferencesInBody(entry.rendered.html, entry.filePath)
+			: entry?.rendered?.html;
+
+	const Content = createComponent(() => serverRender`${unescapeHTML(html)}`);
```

Comment:
nit: it does seem to be an overkill to create a component factory if `html` is `undefined`

Reply:
What would be the right approach for an empty component?